### PR TITLE
[tflite] Java binding for fp16 in NNAPI delegate

### PR DESCRIPTION
**Pull request by freedomtan**
Add Java binding to use the newly added `allow_fp16` in NNAPI delegate.
This is a follow-up of https://github.com/tensorflow/tensorflow/pull/39444

### CODE REVIEW

**Comment by freedomtan** [comment]
tested with something like
```diff
diff --git a/tensorflow/lite/java/demo/app/src/main/java/com/example/android/tflitecamerademo/ImageClassifier.java b/tensorflow/lite/java/demo/app/src/main/java/com/example/android/tflitecamerademo/ImageClassifier.java
index 2e483d8921..e44940dfac 100644
--- a/tensorflow/lite/java/demo/app/src/main/java/com/example/android/tflitecamerademo/ImageClassifier.java
+++ b/tensorflow/lite/java/demo/app/src/main/java/com/example/android/tflitecamerademo/ImageClassifier.java
@@ -183,7 +183,9 @@ public abstract class ImageClassifier {
   }
 
   public void useNNAPI() {
-    nnapiDelegate = new NnApiDelegate();
+    NnApiDelegate.Options options = new NnApiDelegate.Options();
+    options.setAllowFp16(true);
+    nnapiDelegate = new NnApiDelegate(options);
     tfliteOptions.addDelegate(nnapiDelegate);
     recreateInterpreter();
   }
```

**Comment by daverim** [comment]
Hi jdduke@, can you take a look, I'm not too sure about nnapi related work.

File: tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java [hunk 1]
```diff
@@ -118,12 +118,24 @@ public Options setUseNnapiCpu(boolean enable) {
       return this;
     }
 
+    /**
+     * Enable or disable to allow fp32 computation to be run in fp16 in NNAPI.
+     * See https://source.android.com/devices/neural-networks#android-9
+     *
+     * <p>Only effective on Android 9 (API level 28) and above.
+     */
+    public Options setAllowFp16(boolean enable) {
```
Comment (jdduke) [inline tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java:127]: Can you also add a @Deprecated annotation to Interpreter.Options.setAllowFp16PrecisionForFP32 and point them to this?

**Review by jdduke — APPROVED** [review]

**Review by rthadur — APPROVED** [review]

File: tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java (hunk 1 above)
Comment (rthadur) [reply tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java:127]: @freedomtan can you please add a @depricated

**Review by rthadur — CHANGES_REQUESTED** [review]

File: tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java (hunk 1 above)
Comment (freedomtan) [reply tensorflow/lite/delegates/nnapi/java/src/main/java/org/tensorflow/lite/nnapi/NnApiDelegate.java:127]: @rthadur this is the newly added one. I added `@deprecated` to `tensorflow/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java` as suggested by @jdduke. 

Original: set allow_fp16 option is associated with the Interpreter. This is considered deprecated.
change: an option added to the NNAPI delegate code, this is not deprecated.

> @freedomtan can you please add a @depricated

**Review by rthadur — DISMISSED** [review]

File: tensorflow/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java [hunk 2]
```diff
@@ -102,7 +102,9 @@ public Options setUseNNAPI(boolean useNNAPI) {
      * Sets whether to allow float16 precision for FP32 calculation when possible. Defaults to false
      * (disallow).
      *
-     * <p>WARNING: This is an experimental API and subject to change.
+     * @deprecated Prefer using
+     * {@link org.tensorflow.lite.nnapi.NnApiDelegate.Options#setAllowFp16(boolean enable)}.
+     *
      */
     public Options setAllowFp16PrecisionForFp32(boolean allow) {
```
Comment (jdduke) [inline tensorflow/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java:109]: You can add the literal @Deprecated annotation here, in code. See also setUseNNAPI and setNumThreads on the Interpreter class.

**Review by jdduke — CHANGES_REQUESTED** [review]

File: tensorflow/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java (hunk 2 above)
Comment (freedomtan) [reply tensorflow/lite/java/src/main/java/org/tensorflow/lite/Interpreter.java:109]: Argh, I see. I added javadoc description but didn't add `@deprecated` to the method. Thanks

**Review by multiverse-tf — DISMISSED** [review]

**Review by jdduke — DISMISSED** [review]

**Comment by rthadur** [comment]
@freedomtan can you please check this error, its blocking internal submission 

`/tensorflow/lite/java/src/test/java/org/tensorflow/lite/InterpreterTest.java:75: warning: [deprecation] setAllowFp16PrecisionForFp32(boolean) in Options has been deprecated
                .setAllowFp16PrecisionForFp32(false)
                ^/tensorflow/lite/java/src/test/java/org/tensorflow/lite/InterpreterTest.java:397: warning: [deprecation] setAllowFp16PrecisionForFp32(boolean) in Options has been deprecated
            new Interpreter.Options().setUseNNAPI(true).setAllowFp16PrecisionForFp32(true));
                                                       ^
error: warnings found and -Werror specified`

**Comment by freedomtan** [comment]
@rthadur Thanks. Yes, I can reproduce the problem with `bazel build --config opt  tensorflow/lite/java:InterpreterTest`. Should I simply add ` @SuppressWarnings("deprecation")` to ignore those two warnings or rewrite the two tests?

**Comment by rthadur** [comment]
@freedomtan I believe you can add `@SuppressWarnings("deprecation")` , @jdduke is that fine ?

**Comment by freedomtan** [comment]
@jdduke suppress warnings in `//tensorflow/lite/java:InterpreterTest` and add an `allowFp16` test to `//tensorflow/lite/java:NnApiDelegateTest`. Tested them on my desktop Linux machine. Dunno how to run tests on real android devices though.

**Review by rthadur — APPROVED** [review]

**Pull request merged**